### PR TITLE
BE: Messaging

### LINKED ISSUE #3: BE: Messaging
Might require multiple models (conversation, messages)
Routes:
- Create new conversation
- Send message
- Get all conversations
- Get all messages for a conversation

### PR DESCRIPTION
**Pull request by navjot-chahal**
### What this PR does:
- Adds route handlers for `/message`, `/conversation` and `/conversation/all`.
- Adds controllers for messaging route handlers.
- Adds models for `message` and `conversation` collections.

### Screenshots of manual test with Postman:

- Three test users i.e. `test`, `test2`, and `test3` were populated in DB.
- User `test` is the logged-in user for the results.
- Following are the results of manual tests for API endpoints using Postman.

- POST /conversation

Post conversation

- POST /message

Post message

- GET /conversation 

Get conversation

- GET /conversation/all

Get all conversations

### Any information needed to test this feature:
- Messaging BE was tested manually with Postman and is not recommended for future tests.
- A dedicated mocha integration test should be made for quicker testing.

closes #3

### CODE REVIEW

File: server/controllers/conversation.js
Comment (sundayezeilo) [inline server/controllers/conversation.js]: Good documentation

File: server/controllers/conversation.js
Comment (sundayezeilo) [inline server/controllers/conversation.js]: I think you should validate this and return a bad request (`400`) if it's not supplied by the client. It's different from when supplied but no conversation with that is found.

File: server/controllers/conversation.js
Comment (sundayezeilo) [inline server/controllers/conversation.js:26]: :+1:

File: server/controllers/conversation.js
Comment (sundayezeilo) [inline server/controllers/conversation.js]: break `.populate` into the next line for proper formatting

File: server/controllers/conversation.js
Comment (sundayezeilo) [inline server/controllers/conversation.js:48]: :+1:

File: server/controllers/conversation.js
Comment (sundayezeilo) [inline server/controllers/conversation.js:50]: Also validate the request body here

File: server/controllers/conversation.js
Comment (sundayezeilo) [inline server/controllers/conversation.js:62]: Love this query!

File: server/controllers/conversation.js
Comment (sundayezeilo) [inline server/controllers/conversation.js:66]: This should be a `400`. Might be a typo :smile:

File: server/controllers/conversation.js
Comment (sundayezeilo) [inline server/controllers/conversation.js:76]: :+1:

File: server/controllers/conversation.js
Comment (sundayezeilo) [inline server/controllers/conversation.js]: I think `execPopulate` has been deprecated in recent versions of mongoose. Maybe just `exec` should just work fine.

File: server/controllers/conversation.js
Comment (sundayezeilo) [inline server/controllers/conversation.js]: :+1:

File: server/controllers/conversation.js
Comment (sundayezeilo) [inline server/controllers/conversation.js:103]: The `find` method always returns an array. So this check will always be falsy since even an empty array is truth. I think you should return the result to the client even if it's an empty array, so the client should figure that out in rendering.

File: server/controllers/conversation.js
Comment (sundayezeilo) [inline server/controllers/conversation.js:109]: :+1:

File: server/controllers/message.js
Comment (sundayezeilo) [inline server/controllers/message.js:7]: :+1:

File: server/controllers/message.js
Comment (sundayezeilo) [inline server/controllers/message.js:9]: Need to validate these ones

File: server/controllers/message.js
Comment (sundayezeilo) [inline server/controllers/message.js:22]: :+1:

File: server/controllers/message.js
Comment (sundayezeilo) [inline server/controllers/message.js]: Nice one!

File: server/models/Conversation.js
Comment (sundayezeilo) [inline server/models/Conversation.js:20]: :+1:

File: server/models/Message.js
Comment (sundayezeilo) [inline server/models/Message.js:20]: :+1:

File: server/routes/conversation.js
Comment (sundayezeilo) [inline server/routes/conversation.js]: Since your base URL is already `'/conversations'`, you could omit this `all` and just go with `router.route("/")`

**Review by sundayezeilo — COMMENTED** [review]
Awesome! Well-organised and well-implemented. Clean PR. :+1: 
Just a few minor comments to check out so it will be approved.
Happy coding!

File: server/controllers/conversation.js
Comment (navjot-chahal) [reply server/controllers/conversation.js:66]: Hi @ezeilo-su, this is not a typo. I specifically added `409` error based on a [thread](https://stackoverflow.com/questions/9269040/which-http-response-code-for-this-email-is-already-registered/9270432) mentioned by @ugur-akin in previous PR #63. I think already having a conversation in DB is a conflict, and should fall under `res 409`.

File: server/routes/conversation.js
Comment (navjot-chahal) [reply server/routes/conversation.js]: Hi @ezeilo-su, can you please expand on your concern. As per my implementation, I have the GET `/conversation/` route already set up to return the messages for the `conversationId` parsed from the req body. GET `/conversation/all` will return all the conversations for the currently logged-in users instead. Omitting the`/all` route will add a conflict between them.

Comment (sundayezeilo) [reply server/routes/conversation.js]: @navjot-chahal, I only meant that you could omit the `all` phrase there if you wanted to. That's a matter of choice, though.

Comment (navjot-chahal) [reply server/routes/conversation.js]: My apologies @ezeilo-su but, I still did not understand the issue here. If we remove the `/all` from `getAllConversations` route then we would have two conflicting routes (See highlighted image attached). How will the Express-router know where to traffic incoming GET `/conversation/` when it sees two similar routes? Is there anything I am missing?😅
image

Comment (sundayezeilo) [reply server/routes/conversation.js]: Oh, my bad! @navjot-chahal, I'm very sorry I didn't factor in the first route in the comment. I think for the route you use to get conversation based on `conversationId`, that should be something like `router.route("/:conversationId").get(protect, getConversation);`  Then `conversationId` will be available in `req.params.conversationId`. That should be fine. Please reach out to me on Slack if you still need further clarification :pray:

File: server/controllers/conversation.js
Comment (navjot-chahal) [reply server/controllers/conversation.js]: Fixed in next commit.
image

File: server/controllers/conversation.js
Comment (navjot-chahal) [reply server/controllers/conversation.js]: @ezeilo-su Our team recently agreed on a set of prettier settings and this is part of that. I understand this will look better, but overwriting this will result in inconsistency with my colleagues. If another method is called on this chain then prettier will automatically format to the next line as shown in the example below.

image

File: server/controllers/conversation.js
Comment (navjot-chahal) [reply server/controllers/conversation.js:103]: Awesome!  I should have added if (!allConversations || !allConversation.length) instead. Yes, I agree we should send an empty array anyway so I will remove this check.

**Comment by navjot-chahal** [comment]
Hi @ezeilo-su, thanks for your comments on PR. See commit `Hotfix: added Validators` for the requested changes. Also, I just realized I could have used a better approach of `Express-validator` instead 😅. I have the `messaging integration` ticket as well, so maybe I will add that in there. Every day we learn more and get better at things, haha 🤓.

File: server/controllers/conversation.js
Comment (rajivtitus) [inline server/controllers/conversation.js]:
```suggestion
  }).populate("sendByUser", "username email");
```

File: server/controllers/conversation.js
Comment (rajivtitus) [reply server/controllers/conversation.js:66]: @navjot-chahal Yes, the 409 status code is appropriate in this case 👍🏽

File: server/controllers/conversation.js
Comment (rajivtitus) [reply server/controllers/conversation.js:103]: @navjot-chahal please remove this check here as we don't really need it 😄

File: server/models/Conversation.js
Comment (rajivtitus) [inline server/models/Conversation.js:7]:
```suggestion
      ref: "User",
```
Always use PascalCase for model names. However, I think changing it at this point may cause the app to break (so you can refrain from doing so) 😄

File: server/models/Conversation.js
Comment (rajivtitus) [inline server/models/Conversation.js:20]: 👍🏽

File: server/controllers/message.js
Comment (rajivtitus) [inline server/controllers/message.js]: Did you intend to push the entire document? Based on the model below it looks like we only need the 'id' of the message

File: server/models/Message.js
Comment (rajivtitus) [inline server/models/Message.js]:
```suggestion
    sender: {
```

File: server/routes/conversation.js
Comment (rajivtitus) [inline server/routes/conversation.js]:
```suggestion
router.route("/:conversationId").get(protect, getConversation);
```
I would recommend using path parameters are we are fetching a specific resource here :)

File: server/routes/conversation.js
Comment (rajivtitus) [inline server/routes/conversation.js]:
```suggestion
router.route("/").get(protect, getAllConversations);
```

File: server/controllers/conversation.js
Comment (rajivtitus) [inline server/controllers/conversation.js]: GET requests don't have a request body :)

File: server/controllers/conversation.js
Comment (rajivtitus) [inline server/controllers/conversation.js]: 💯

**Review by rajivtitus — COMMENTED** [review]
Impressive work here 🚀 

Just some minor comments for you to address :)

File: server/controllers/conversation.js
Comment (sundayezeilo) [inline server/controllers/conversation.js:81]: :+1:

File: server/controllers/conversation.js
Comment (sundayezeilo) [inline server/controllers/conversation.js:54]: :+1:

**Review by sundayezeilo — COMMENTED** [review]
Nice job implementing those changes :+1: Please also resolve the merge conflict.
Happy coding!

File: server/routes/conversation.js
Comment (navjot-chahal) [reply server/routes/conversation.js]: @rajivtitus Removing this will cause conflict among routes. There is already a GET route at "/conversation/" which gets all messages. 🙂

File: server/controllers/conversation.js
Comment (navjot-chahal) [reply server/controllers/conversation.js]: Actually, I tested it and this does not work 😅. `conversation.firstUser` is an object which I am trying to strictly compare to a string. I will cast this to String or use an abstract comparison operator instead (`==`).

File: server/controllers/conversation.js
Comment (navjot-chahal) [reply server/controllers/conversation.js]: @rajivtitus I checked W3's specs for HTTP/1.1, the GET method can have the body as per my understanding. Also, the request is handled properly in the test with a proper response. Do you mean to just use URI for better practice and consistency? I will add a URL string-based request with the next commit. 😄

File: server/routes/conversation.js
Comment (navjot-chahal) [reply server/routes/conversation.js]: I understand this change will work after making the above changes.

File: server/models/Conversation.js
Comment (navjot-chahal) [reply server/models/Conversation.js:7]: Oh, I was just following the standard from the previous model. But if you insist I can use PascalCase from now onwards. 🙂

File: server/controllers/message.js
Comment (navjot-chahal) [reply server/controllers/message.js]: Yes, you are absolutely right. It should be `lastMessage._id`. I am surprised it passed my test as moongoose/MongoDB is smart enough to add only the id automatically.

**Comment by navjot-chahal** [comment]
@rajivtitus I added the requested changes in commit `Hotfix - 2` and here are the test results for the modified endpoints.

- POST /message

Post message

- GET /conversation

Get all conversation

- GET /conversation/:conversationId

Get conversation messages

File: server/controllers/conversation.js
Comment (sundayezeilo) [inline server/controllers/conversation.js:5]: :+1:

File: server/controllers/conversation.js
Comment (sundayezeilo) [inline server/controllers/conversation.js:9]: That's the most popular practice :smiley: :

File: server/models/Message.js
Comment (sundayezeilo) [inline server/models/Message.js:5]: :+1:

File: server/routes/conversation.js
Comment (sundayezeilo) [inline server/routes/conversation.js:10]: :+1:

File: server/routes/conversation.js
Comment (sundayezeilo) [inline server/routes/conversation.js:12]: :+1:

File: server/routes/conversation.js
Comment (sundayezeilo) [inline server/routes/conversation.js:14]: :+1:

**Review by sundayezeilo — COMMENTED** [review]
I'm glad you made those changes :smile:
Please just resolve the merge conflict and wait for @rajivtitus's review before approval.
Happy coding!

File: server/controllers/conversation.js
Comment (rajivtitus) [reply server/controllers/conversation.js]: @navjot-chahal Sending body/payload in a GET request may cause some existing implementations to reject the request — while not prohibited by the specification, the semantics are undefined. It is better to just avoid sending payloads in GET requests.

**Review by rajivtitus — APPROVED** [review]
Great work making those changes and testing everything once again! 🚀

**Pull request merged**
